--[[
	Hello there!
	Setting up the board is simple.
	
	1. Place the model 'Boards' in the WORKSPACE and don't change it's name.
	2. Do not delete any scripts as it will break the Boards.
	3. You can have as many Boards and Button Boards as you wish. Just clone the 'Buttons' part or 'Screen' part into this model.
	4. Open the script Products and read the instructions.
	5. Enjoy!
	
	NOTE:
	You can change some board settings by going in-game and clicking the cog on the right bottom of the board.
	
	Please Note that the board only works in Online mode, not studio or play solo.
	
	Roblox keeps the donated rs for 3 days before handing it over to you.
	You can view this under the 'pending' column in your summary.
	
	If you have any questions contact Yeah_Ember.
--]]